### DISCUSSION TITLE
Mapping a request onto a data object is not working

### DISCUSSION THREAD

**Discussion by ashishimple** [spatie/laravel-data#46] [Q&A]
it give synatx error

**Comment by rubenvanassche** [comment]
Could you give us a bit more information about what's going wrong exactly?

**Comment by ashishimple** [comment]
@rubenvanassche 

1) Firstly if i use this syntax 
 
```
public static function fromRequest(Request $request): static
 {
        return new self(
            {$request->input('title_of_song')}, 
            {$request->input('artist_name')}
        );
    }
```

it gives syntax error on " { " in the line  {$request->input('title_of_song')}, 

2) my query is that if i have a request coming POST API from frontend like

```
{
	authorInfo: {
		authorName: 'MJ',
		authorLocation: 'USA'
	},
	AlbumInfo: {
		albumName: 'Thriller',
		songs: 15
	}

}
```

then how should i map the coming request , as my database fields with this request with my Data Files ,as my database contains two tables fields mainly
Authors : id, name, location
album: id, author_id, name, no_of_songs


i have tried using nested example i have 3 Data files ,ie. **MainData, AuthorInfoData, AlbumInfoData**

**MainData**

```
public function __construct(
        public AuthorInfoData $authorInfo,
        public AlbumInfoData $albumInfo,
    ) {

    }
```

**and inside AuthorInfoData**

```
public function __construct(
        public ?string $name, 
        public ?string $location,
    ) {

    }

public static function fromRequest(Request $request): static
    {
        return new self(
            {$request->input('authorInfo.authorName')}, 
            {$request->input('authorInfo.authorlocation')}
        );
    }
```


it gives syntax error on  {$request->input('authorInfo.authorName')}, 

and then i manage to change the way of writing the above function like this 

```
public static function fromRequest(Request $request): static
{
        return new self(
            $request->input('authorInfo.authorName'), 
            $request->input('authorInfo.authorlocation')
        );
    }

```

But still it doesnot work

Can u please throw some light on this ? how to manage and map request coming from front end on above scenario

**Comment by rubenvanassche** [comment]
That's because this syntax isn't valid:

```php
public static function fromRequest(Request $request): static
 {
        return new self(
            {$request->input('title_of_song')}, 
            {$request->input('artist_name')}
        );
    }
```

You should do the following:

```php
public static function fromRequest(Request $request): static
 {
        return new self(
            $request->input('title_of_song'), 
            $request->input('artist_name')
        );
    }
```

I've updated the docs on this because they weren't correct on this matter

**Comment by gauravmak** [comment]
Question: Do we need to manually call `DataClass::fromRequest($request)` to take benefit of the mapping feature or is it automatically used even when we resolve the data class from the container?

**Reply by rubenvanassche** [reply]
When we're mapping a request onto a data object when resolving in the container, the package will call the `from` method on the data object. So if a `fromRequest` method exists that one will be called(if the request can be correctly injected). When no such method exists the `fromArray` method will be called which will use an array version of the request.

So to answer your question, yes we'll map as best as possible when resolving data from the container.

**Reply by gauravmak** [reply]
Thanks for the detailed reply @rubenvanassche . I appreciate it.

I had first tried to get an answer by looking into the source code but couldn't find a method call. Then, found [this test](https://github.com/spatie/laravel-data/blob/29cc4bd5f1bfa2c778165051e704c37bbff109f5/tests/RequestDataTest.php#L225).

The other issue that we are facing is:
- The request inputs are in the snake case.
- Respective properties of the data class should be in the camel case.
- But, we have to keep both to make this work. (Test ref - https://github.com/spatie/laravel-data/blob/29cc4bd5f1bfa2c778165051e704c37bbff109f5/tests/RequestDataTest.php#L241)

Is there any workaround?

PS - The new design of Ignition is beautiful. Great work 👏

**Reply by rubenvanassche** [reply]
Hi @gauravmak,

At the moment the mapping of properties is quite a problem a lot of people have, since sometimes properties are coming in as snakecase and are data properties in camelcase and vice versa. We're planning to fix this in V2 with specific attributes and mappers.

As for the case you describe with the test, laravel-data v1.4.2 which will be released later today will include a WithoutValidation attribute which will not try the validate the property when the data is coming from a request.
